keps/sig-autoscaling/draft-20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,170 @@
+[Vertical Resources Scaling in Kubernetes][].
+
+[proposal for live and in-place vertical scaling]: https://github.com/kubernetes/community/pull/1719
+[Vertical Resources Scaling in Kubernetes]: https://docs.google.com/document/d/18K-bl1EVsmJ04xeRq9o_vfY2GDgek6B6wmLjXw-kos4/edit?ts=5b96bf40
+
+## Motivation
+
+Resources allocated to a Pod's Container can require a change for various reasons:
+* load handled by the Pod has increased significantly and current resources are
+  not enough to handle it,
+* load has decreased significantly and currently allocated resources are unused
+  and thus wasted,
+* Resources have simply been set improperly.
+
+Currently changing Resources allocation requires the Pod to be recreated since
+the PodSpec is immutable.
+
+While many stateless workloads are designed to withstand such a disruption, some
+are more sensitive, especially when using low number of Pod replicas.
+
+Moreover, for stateful or batch workloads, a Pod restart is a serious
+disruption, resulting in lower availability or higher cost of running.
+
+Allowing Resources to be changed without recreating a Pod nor restarting a
+Container addresses this issue directly.
+
+### Goals
+
+* Primary: allow to change Pod resource requests & limits without restarting its
+  Containers.
+* Secondary: allow actors (users, VPA, StatefulSet, JobController) to decide
+  how to proceed if in-place resource update is not available.
+* Secondary: allow users to specify which Pods and Containers can be updated
+  without a restart.
+
+### Non-Goals
+
+The explicit non-goal of this KEP is to avoid controlling full life-cycle of a
+Pod which failed an in-place resource update. These cases should be handled by
+actors which initiated the update.
+
+Other identified non-goals are:
+* allow to change Pod QoS class without a restart,
+* to change resources of Init Containers without a restart.
+
+## Proposal
+
+### API Changes
+
+PodSpec becomes mutable with regards to resources and limits.
+Additionally, PodSpec becomes a Pod subresource to allow fine-grained access control.
+
+PodStatus is extended with information about actually allocated resources.
+
+Thanks to the above:
+* PodSpec.Container.ResourceRequirements becomes purely a declaration,
+  denoting **desired** state of the Pod,
+* PodStatus.ContainerStatus.ResourceAllocated (new object) denotes **actual**
+  state of the Pod resources.
+
+To distinguish between possible states of the Pod resources,
+a new PodCondition InPlaceResize is added, with the following states:
+* (empty) - the default value; resource update awaits reconciliation
+  (if ResourceRequirements differs from ResourceAllocated),
+* Awaiting - awaiting resources to be freed (e.g. via pre-emption),
+* Failed - resource update could not have been performed in-place
+  but might be possible if some conditions change,
+* Rejected - resource update was rejected by any of the components involved.
```

Comment:
What components are consulted? Could you add some examples of why an update could be rejected, and how the control flow might look?

Reply:
In our design, scheduler decides whether in-place update is possible. One reason it can fail is if scheduler's view of node available resources says the node can't currently accommodate the request. Another reason it can fail is if policy allows scheduler to reschedule the pod to resize, and reschedule action results in PDB violation. (Such failures are handled through retries in our proposal)

In multiple scheduler case, kubelet can also fail a pod resources update request due to race conditions, say if a second scheduler binds a pod to the same node, and is handled before the update request lands on the node.

I'll reference our diagram in the 'Vertical Resources Resizing Workflow' section of our [design doc](https://docs.google.com/document/d/18K-bl1EVsmJ04xeRq9o_vfY2GDgek6B6wmLjXw-kos4),  but the implementation here might evolve in a different direction.

Reply:
Thanks for linking the detailed design. The reason I asked is because of corner cases for 1) CPU affinity and 2) device resources. In (1) some applications may make their own thread-level affinity decisions after starting, and the policy provided by the CPU manager explicitly promises never to change the assigned cpuset for the lifetime of the container. In (2) there are likely considerations for applications that consume accelerator devices; usually they configure themselves based on available hardware at initialization time and changing the device whitelist after that could cause problems.

Both are surmountable, but given that these features already have users they should be taken into account somehow in this proposal. One idea based on the design doc could be an additional resize policy like ResizePolicyNo to opt out.

Reply:
Yes both cases seem to be good candidates for opt-out, which is currently supported by VPA 'Off' mode which still allows the user to see VPA recommendations for later analysis.

Reply:
Since the in-place update might fail or be rejected, how to manage users' expectation? Users will never know these things in advance: whether the in-place update will success or not, how many pods will be in-place updated or be rescheduled. Some users could not make a decision without it, e.g. they want in-place update because they do not want Pod ip be changed, if lots of Pod's IPs might be changed, they'd rather not update Pods.

Reply:
> Since the in-place update might fail or be rejected, how to manage users' expectation? Users will never know these things in advance: whether the in-place update will success or not, how many pods will be in-place updated or be rescheduled. Some users could not make a decision without it, e.g. they want in-place update because they do not want Pod ip be changed, if lots of Pod's IPs might be changed, they'd rather not update Pods.

This is controlled by a set of policies. Referencing our [earlier suggestion](https://github.com/kubernetes/community/pull/2908/commits/4ad6fa7c27f4a21c27a6be83c2dc81c43549fa55#r238145945), our proposal is to have two levels of policy controls. Please review that and see if it would work for the cases you have in mind.

Our implementation respects PDBs if reschedule is allowed by policy, and in-place is not possible. For failures, our approach uses retries from the controller that owns the pods. VPA can also handle failures directly with retry disabled.

Reply:
While the Control Flow of this KEP is still to be filled in, one of the ideas we had was that:

1. an update to ResourceRequirements is validated first (e.g. an update changing QoS class is rejected immediately),
1. then AdmissionControllers kick-in, for example verifying Quota,
1. then Scheduler checks if the Pod first, possibly pre-empting lower-priority Pods from the same Node if it's required,
1. at the same time Kubelet tries to apply the new ResourceRequirements; if this happens prior to a required pre-emption, a retry by Kubelet will be required once Scheduler finished pre-emption and clears InPlaceResize condition.

As for handling applications that assume that some conditions present at init time remain constant, we've foreseen the ResizingPolicy flag (see further in the proposal).
It allows to force a Container or Pod restart, depending on the needs of your application.
@ConnorDoyle , please confirm whether this would be enough for the use-cases you've mentioned (CPU affinity and device resources).

@hex108 - as explained by @vinaykul the first option is to use PDB to control the rate of updates. One possibilities would be for in-place update to temporarily consume PDB while an update is in progress and either give it back (on success) or consume permanently (on failure causing a Pod restart).
There is also an option of adding a Pod-level policy (as proposed in the design mentioned by @vinaykul ) to let the users forbid in-place actions for given Pods. I'm not yet 100% sure if it would be better to add it here or in the controllers that start in-place updates or what all the combinations with ResizingPolicy (container-level) flag can mean but that's a discussion we should definitely have.

Reply:
@kgolab If we have scheduler and kubelet process the resource update at the same time (steps 3 & 4), we might end up in a race that could lead to unpredictable bouncing around. e.g. if scheduler assigned a new pod to that node and used up capacity just before it saw the resource update request, then it will NAK the update request which the kubelet accepted in parallel. then kubelet sees the new pod assignment and rejects it, leading the scheduler to have to go through the high-cost scheduling process again, and update request to be retried as well. 

IMHO, it is better to serialize the flow through the scheduler. if scheduler agrees to the update request, it updates its cache and updates PodStatus ResourceAllocated. kubelet only looks at what ResourceAllocated says and applies that. this also handles the multiple scheduler race condition well. Thoughts?

Reply:
+1 to @vinaykul's point and approach. I also think that we should try to limit races as much as possible.

Reply:
We have updated the design to not have Scheduler in the resize path at this time. VPA mode 'Off' accounts for the resize opt-out request above. I'm resolving this comment. Please unresolve if anything is still not addressed.